### PR TITLE
Feature Toggles: Create API for updating feature toggle state from the feature toggle admin page 

### LINKED ISSUE #70926: Apply feature toggle filters based on feature state
We should have all experimental toggles hidden by default, and all privatePreview toggles read-only by default, to reduce the likelihood of a flaky toggle breaking our instances.

### LINKED ISSUE #72607: Feature Toggle Admin Page: Implement backend API for POSTing feature toggle state
**POST /api/featuremgmt/toggle** (for Phase 1, this will just be a stub)
Accepts a request to toggle a feature toggle and, if allowed, sends a signal to the HG infrastructure to update the toggle and restart the instance
Sample payload:
```
{
	"toggle": "useCachingService",
	"enable": true
}

```
Endpoint will be guarded by access control middleware to ensure that only users with appropriate privileges can view and update feature toggles.

### PR DESCRIPTION
**Pull request by mmandrus**
**What is this feature?**

This adds all the scaffolding needed to update feature toggle status from the feature toggle admin page. The update endpoint still does not send a request off to Hosted Grafana, but we have put all the necessary guards in place. 

In addition, this PR achieves the following:

- Makes Unknown, Experimental, and Private Preview toggles hidden by default
- Makes Public Preview toggles read-only by default
- Disallows editing the toggle that controls this feature
- Various refactors to clean up increasingly unwieldy code and tests

**Why do we need this feature?**

Eventually we want to allow Grafana Cloud users to be able to update GA feature toggles themselves without going through support. We still have to determine what this will look like on the HG-side, but we can build everything in the meantime.

Note that while the update function is unimplemented, it is still behind an experimental feature toggle and thus should be safe to merge what we have for now.

**Who is this feature for?**

Grafana operators and users wishing to make feature toggle management easier

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/72607
Fixes https://github.com/grafana/grafana/issues/70926

### CODE REVIEW

**Review by leandro-deveikis — APPROVED** [review]
LGTM!

**Review by IbrahimCSAE — APPROVED** [review]
LGTM

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 1]
```diff
@@ -2257,6 +2257,34 @@ For more information about Grafana Enterprise, refer to [Grafana Enterprise]({{<
...
+
+## [feature_management]
+
+The options in this section configure the experimental Feature Toggle Admin Page feature, which is enabled using the `featureToggleAdminPage` feature toggle. **Use at your own risk.**
+
+Please see [Configure feature toggles]({{< relref "/feature-toggles" >}}) for more information.
+
+### allow_editing
+
+Allow users to toggle feature toggle state in the feature management page. Default is `false`.
```
Comment (JStickler) [inline docs/sources/setup-grafana/configure-grafana/_index.md:2270]:
```suggestion
Lets you switch the feature toggle state in the feature management page. The default is `false`.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 2]
```diff
@@ -2257,6 +2257,34 @@ For more information about Grafana Enterprise, refer to [Grafana Enterprise]({{<
...
+
+### update_controller_url
+
+Set the URL of the controller that manages feature toggle updates. If not set, feature toggles in the feature management page will be read-only.
```
Comment (JStickler) [inline docs/sources/setup-grafana/configure-grafana/_index.md:2274]:
```suggestion
Set the URL of the controller that manages the feature toggle updates. If not set, feature toggles in the feature management page will be read-only.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 3]
```diff
@@ -2257,6 +2257,34 @@ For more information about Grafana Enterprise, refer to [Grafana Enterprise]({{<
...
+
+**Note:** The API for feature toggle updates has not been defined yet; please check back later.
```
Comment (JStickler) [inline docs/sources/setup-grafana/configure-grafana/_index.md:2276]:
```suggestion
{{% admonition type="note" %}}
The API for feature toggle updates has not been defined yet.
{{% /admonition %}}
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 4]
```diff
@@ -2257,6 +2257,34 @@ For more information about Grafana Enterprise, refer to [Grafana Enterprise]({{<
...
+
+### hidden_toggles
+
+Hide additional specific feature toggles from the feature management page. By default, feature toggles in stages `unknown`, `experimental`, and `private preview` are hidden from the UI. Use this option to hide toggles in stages `public preview`, `general availability`, and `deprecated`.
```
Comment (JStickler) [inline docs/sources/setup-grafana/configure-grafana/_index.md:2280]:
```suggestion
Hide additional specific feature toggles from the feature management page. By default, feature toggles in the `unknown`, `experimental`, and `private preview` stages are hidden from the UI. Use this option to hide toggles in the `public preview`, `general availability`, and `deprecated` stages.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 5]
```diff
@@ -2257,6 +2257,34 @@ For more information about Grafana Enterprise, refer to [Grafana Enterprise]({{<
...
+
+### read_only_toggles
+
+Disable updates for additional specific feature toggles in the feature management page. By default, feature feature toggles can only be updated if they are in stages `general availability` and `deprecated`. Use this option to disable updates for toggles in those stages.
```
Comment (JStickler) [inline docs/sources/setup-grafana/configure-grafana/_index.md:2284]:
```suggestion
Use to disable updates for additional specific feature toggles in the feature management page. By default, feature toggles can only be updated if they are in the `general availability` and `deprecated`stages. Use this option to disable updates for toggles in those stages.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 6]
```diff
@@ -2257,6 +2257,34 @@ For more information about Grafana Enterprise, refer to [Grafana Enterprise]({{<
 
 Keys of alpha features to enable, separated by space.
 
+<hr>
+
+## [feature_management]
+
+The options in this section configure the experimental Feature Toggle Admin Page feature, which is enabled using the `featureToggleAdminPage` feature toggle. **Use at your own risk.**
```
Comment (JStickler) [inline docs/sources/setup-grafana/configure-grafana/_index.md:2264]: I can't find any other places in the docs were we use the phrase "Use at your own risk."  Currently consulting with the team on if we have a standard caution for experimental features.

**Review by JStickler — COMMENTED** [review]
[docs team]

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 6 above)
Comment (JStickler) [reply docs/sources/setup-grafana/configure-grafana/_index.md:2264]:
```suggestion
The options in this section configure the experimental Feature Toggle Admin Page feature, which is enabled using the `featureToggleAdminPage` feature toggle. Projects and features in the experimental stage are not meant to be used in production environments, and the risks are unknown.
```

Comment (JStickler) [reply docs/sources/setup-grafana/configure-grafana/_index.md:2264]: Another option:
```suggestion
The options in this section configure the experimental Feature Toggle Admin Page feature, which is enabled using the `featureToggleAdminPage` feature toggle. Grafana Labs offers support on a best-effort basis, and breaking changes might occur prior to the feature being made generally available.
```

**Review by JStickler — APPROVED** [review]
[docs team] LGTM

File: pkg/api/featuremgmt.go [hunk 7]
```diff
@@ -1,30 +1,95 @@
...
+	cmd := featuremgmt.UpdateFeatureTogglesCommand{}
+	if err := web.Bind(ctx.Req, &cmd); err != nil {
+		return response.Error(http.StatusBadRequest, "bad request data", err)
+	}
+
+	for _, t := range cmd.FeatureToggles {
+		// make sure flag exists, and only continue if flag is writeable
+		if f, ok := hs.Features.LookupFlag(t.Name); ok && isFeatureWriteable(f, hs.Cfg.FeatureManagement.ReadOnlyToggles) {
+			hs.log.Info("UpdateFeatureToggle: updating toggle", "toggle_name", t.Name, "enabled", t.Enabled, "username", ctx.SignedInUser.Login)
+			// TODO build payload
```
Comment (ArturWierzbicki) [inline pkg/api/featuremgmt.go:58]: how will this work in OSS? is this essentially a cloud-only feature?

**Review by jcalisto — APPROVED** [review]

File: pkg/api/featuremgmt.go (hunk 7 above)
Comment (mmandrus) [reply pkg/api/featuremgmt.go:58]: Hey @ArturWierzbicki, our goal is to make this as widely adoptable as we can. At the very least, it needs to be useable by AWS and Azure managed Grafana instances, but we would like for it to be available for other Grafana operators as well. For what it's worth, we haven't done the phase 2 design yet, so the stuff we have in here for sending the request to a controller URL is still not definite.

One thing to consider is that someone running this in a small-scale environment on-prem probably won't have as much use for this, as they can just edit the ini file and restart. The target audience is more for larger orgs running Grafana as a service.

**Pull request merged**